hack/lib/dev-env.sh

```diff
@@ -0,0 +1,177 @@
+
+# Check that required binaries are installed
+command -v make >/dev/null 2>&1 || { echo >&2 "make is required but it's not installed.  Aborting."; exit 1; }
+command -v docker >/dev/null 2>&1 || { echo >&2 "docker is required but it's not installed.  Aborting."; exit 1; }
+command -v kind >/dev/null 2>&1 || { echo >&2 "kind is required but it's not installed.  Aborting."; exit 1; }
+
+if [[ "${REPO_ROOT:-}" = "" ]]; then
+    echo "Error: \$REPO_ROOT is unset."
+    echo "Set \$REPO_ROOT in the sourcing script, e.g.:"
+    echo 'REPO_ROOT="$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )/.." &> /dev/null && pwd )"'
+    exit 1
+fi
+
+OUTPUT="${OUTPUT:-${REPO_ROOT}/_output}"
+
+# Parameters
+
+# Parameters required to be set by caller for dev environment creation only:
+#   AUTHENTICATOR_IMAGE
+#   ADMIN_ARN
+
+# Parameters with sane defaults: 
+AUTHENTICATOR_BIN="${AUTHENTICATOR_BIN:-${OUTPUT}/bin/aws-iam-authenticator}"
+REGION="${REGION:-us-west-2}"
+APISERVER_URL="${APISERVER_URL:-https://127.0.0.1:6443}"
+CLUSTER_NAME="${CLUSTER_NAME:-authenticator-dev-cluster}"
+NETWORK_NAME="${NETWORK_NAME:-authenticator-dev}"
+NETWORK_SUBNET="${NETWORK_SUBNET:-172.30.0.0/16}"
+AUTHENTICATOR_IP="${AUTHENTICATOR_IP:-172.30.0.10}"
+AUTHENTICATOR_PORT="${AUTHENTICATOR_PORT:-21362}"
+
+# Location of templates, config files, mounts
+authenticator_config_template="${REPO_ROOT}/hack/dev/authenticator.yaml"
+kind_config_template="${REPO_ROOT}/hack/dev/env.yaml"
+kind_config_host_dir="${OUTPUT}/dev/kind"
+kind_config="${kind_config_host_dir}/env.yaml"
+authenticator_config_host_dir="${OUTPUT}/dev/authenticator/config"
+authenticator_config="${authenticator_config_host_dir}/authenticator.yaml"
+authenticator_export_host_dir="${OUTPUT}/dev/authenticator/export"
+authenticator_state_host_dir="${OUTPUT}/dev/authenticator/state"
+authenticator_config_dest_dir="/etc/authenticator"
+authenticator_export_dest_dir="/var/authenticator/export"
+authenticator_state_dest_dir="/var/authenticator/state"
+apiserver_config_dest_dir="/etc/kubernetes/authenticator"
+# Kubeconfig used when authenticator loads its mapping configuration from the API server
+authenticator_kubeconfig="${authenticator_config_dest_dir}/authenticator-kubeconfig.yaml"
+# Kubeconfig passed to the apiserver so it can kind its authentication webhook
+authenticator_generate_kubeconfig="${authenticator_export_dest_dir}/apiserver-auth-webhook-kubeconfig.yaml"
+apiserver_auth_webhook_kubeconfig="${apiserver_config_dest_dir}/apiserver-auth-webhook-kubeconfig.yaml"
+host_path_webhook_kubeconfig="${authenticator_export_host_dir}/apiserver-auth-webhook-kubeconfig.yaml"
+# Kubectl kubeconfig template
+kubectl_kubeconfig_template="${REPO_ROOT}/hack/dev/kubeconfig.yaml"
+client_dir="${OUTPUT}/dev/client"
+# Generated kubeconfig for testing
+kubectl_kubeconfig="${client_dir}/kubeconfig.yaml"
+# Admin kubeconfig generated by kind
+kind_kubeconfig="${client_dir}/kind-kubeconfig.yaml"
+
+function create_network() {
```

Comment:
is there a requirement to create a separate virtual network? Why not use the default docker virtual network?

Reply:
Kind already creates/uses its own network, so in order to include the authenticator in that network I needed to do it this way (or run authenticator on the kind node along with the apiserver, etc, but in that case I would have had to do some setup slightly differently, including generating all config/certs up front instead of letting authenticator generate them.  This is because the ca.cert needs to embedded in a webhook kubeconfig and passed to the apiserver on startup).  Also interacting with the apiserver as a standalone docker container is quite easy, and we don't have to interact with kind for debugging... e.g. port-forwarding in order to curl the metrics endpoint, etc.